File: src/edu/mit/csail/cgs/tools/chipchip/AddMSP.java
package edu.mit.csail.cgs.tools.chipchip;

import java.io.*;
import java.util.*;
import java.sql.*;
import java.util.regex.*;
import edu.mit.csail.cgs.utils.NotFoundException;
import edu.mit.csail.cgs.utils.database.*;
import edu.mit.csail.cgs.datasets.general.*;
import edu.mit.csail.cgs.datasets.species.*;
import edu.mit.csail.cgs.datasets.chipchip.*;

/**
 * MSP files are the output of the Young Lab's error model (also known
 * as the Rosetta error model.  
 *
 * Usage:
 * java edu.mit.csail.cgs.tools.chipchip.AddMSP --species "$SC;SGDv1" --analysis "Sc Gcn4 in YPD;linefitting normalization"  -- file1.msp file2.msp
 *
 */
public class AddMSP {

    private Organism species;
    private Genome genome;
    private String analysisname, analysisversion;
    private int analysisid;
    private java.sql.Connection core, chipchip;
    private int chromcol, poscol, ratiocol, xcol, pvalcol, pval3col, redcol, greencol, mediancol;
    private Map<String,Integer> chrommap;
    private ArrayList<String> fnames;

    public AddMSP() throws UnknownRoleException, SQLException {
        core = DatabaseFactory.getConnection("core");
        chipchip = DatabaseFactory.getConnection("chipchip");
        fnames = new ArrayList<String>();
    }

    public void parseArgs(String[] args) throws NotFoundException {
        species = null;
        genome = null;
        analysisname = null;
        analysisversion = null;
        int i;
        for (i = 0; i < args.length; i++) {
            if (args[i].equals("--species")) {
                String s = args[++i];
                String pieces[] = s.split(";");                
                species = new Organism(pieces[0]);
                if (pieces.length >= 2) {
                    genome = species.getGenome(pieces[1]);
                }
            }
            if (args[i].equals("--analysis") ||
                args[i].equals("--analysisname")) {
                String s = args[++i];
                String pieces[] = s.split(";");                
                analysisname = pieces[0];
                if (pieces.length >= 2) {
                    analysisversion = pieces[1];
                }
            }
            if (args[i].equals("--analysisversion")) {
                analysisversion = args[++i];
            }
            if (args[i].equals("--file")) {
                fnames.add(args[++i]);
            }
            if (args[i].equals("--")) {
                break;
            }
        }
        for (;i < args.length; i++) {
            fnames.add(args[i]);
        }

        for (i = 0; i < args.length; i++) {
            if (args[i].equals("--genome")) {
                genome = species.getGenome(args[++i]);
            }
        }
        if (species == null || genome == null) {
            throw new RuntimeException("Must supply --species 'species;genomeversion'");
        }
        if (analysisname == null || analysisversion == null) {
            throw new RuntimeException("Must supply --analysis 'analysisname;analysisversion'");
        }
    }

    /* parseArgs must be called first.  Ensures that the analysis specified
       by analysisname and analysisversion exists and sets analysisid accordingly. */
    public void createAnalysis() throws SQLException {
        Statement stmt = chipchip.createStatement();
        ResultSet rs = stmt.executeQuery("select id from rosettaanalysis where name = '" + 
                                         analysisname + "' and version = '" +
                                         analysisversion + "' and species =" + species.getDBID());
        if (rs.next()) {
            analysisid = rs.getInt(1);
        } else {
            rs.close();
            ArrayList<String> fieldnames = new ArrayList<String>();
            ArrayList<String> values = new ArrayList<String>();
            fieldnames.add("id"); values.add(Sequence.getInsertSQL(chipchip,"analysis_id"));
            fieldnames.add("name"); values.add("'" + analysisname + "'");
            fieldnames.add("version"); values.add("'" + analysisversion + "'");
            fieldnames.add("species"); values.add(Integer.toString(species.getDBID()));
            fieldnames.add("active"); values.add("1");
            Pattern exptname = Pattern.compile("\\s (\\w.*):(.*):(.*) vs (.*):(.*):(.*\\w)");
            Matcher matcher = exptname.matcher(analysisname);            
            if (matcher.matches()) {
                MetadataLoader loader = new MetadataLoader();
                fieldnames.add("factorone");
                values.add(Integer.toString(loader.getFactor(matcher.group(1)).getDBID()));
                fieldnames.add("cellsone");
                values.add(Integer.toString(loader.getCells(matcher.group(2)).getDBID()));
                fieldnames.add("conditionone");
                values.add(Integer.toString(loader.getCondition(matcher.group(3)).getDBID()));

                fieldnames.add("factortwo");
                values.add(Integer.toString(loader.getFactor(matcher.group(4)).getDBID()));
                fieldnames.add("cellstwo");
                values.add(Integer.toString(loader.getCells(matcher.group(5)).getDBID()));
                fieldnames.add("conditiontwo");
                values.add(Integer.toString(loader.getCondition(matcher.group(6)).getDBID()));               
            }

            StringBuffer cmd = new StringBuffer("insert into rosettaanalysis(");
            for (int i = 0; i < fieldnames.size(); i++) {
                cmd.append(fieldnames.get(i));
                if (i < fieldnames.size() -1 ) {
                    cmd.append(", ");
                }
            }
            cmd.append(") values(");
            for (int i = 0; i < values.size(); i++) {
                cmd.append(values.get(i));
                if (i < values.size() -1 ) {
                    cmd.append(", ");
                }
            }
            cmd.append(")");
            

            stmt.executeUpdate(cmd.toString());
            rs = stmt.executeQuery(Sequence.getLastSQLStatement(chipchip,"analysis_id"));
            rs.next();
            analysisid = rs.getInt(1);            
        }
        rs.close();

        rs = stmt.executeQuery("select count(*) from rosettaToGenome where analysis = " + 
                            analysisid + " and genome = " + genome.getDBID());
        rs.next();
        if (rs.getInt(1) == 0) {
            stmt.executeUpdate("insert into rosettaToGenome(analysis, genome) values (" +
                            analysisid + "," + genome.getDBID() + ")");
        }
        rs.close();
        stmt.close();
    }

    public void createChromMap() {
        chrommap = genome.getChromIDMap();
        if (species.getName().equals("Homo sapies")) {
            chrommap.put("23",chrommap.get("X"));
            chrommap.put("24",chrommap.get("Y"));
            chrommap.put("25",chrommap.get("mt"));
        }
        if (species.getName().equals("Mus musculus")) {
            chrommap.put("20",chrommap.get("X"));
            chrommap.put("21",chrommap.get("Y"));
            chrommap.put("22",chrommap.get("mt"));
        }
        if (species.getName().equals("Danio rerio")) {
            chrommap.put("26",chrommap.get("Un"));
            chrommap.put("27",chrommap.get("NA"));
        }
        if (species.getName().equals("Drosophila melanogaster")) {
            chrommap.put("1",chrommap.get("2L"));
            chrommap.put("2",chrommap.get("2R"));
            chrommap.put("3",chrommap.get("3L"));
            chrommap.put("4",chrommap.get("4"));
            chrommap.put("5",chrommap.get("3R"));
            chrommap.put("6",chrommap.get("X"));
            chrommap.put("7",chrommap.get("Y"));
        }

    }

    public void getColumnHeaders(String hline) {
        String colnames[] = hline.split("\\t");
        for (int i = 0; i < colnames.length; i++) {
            colnames[i] = colnames[i].toLowerCase();
            if (colnames[i].equals("chr")) {
                chromcol = i;
            } else if (colnames[i].equals("pos")) {
                poscol = i;
            } else if (colnames[i].equals("ratio")) {
                ratiocol = i;
            } else if (colnames[i].equals("x'")) {
                xcol = i;
            } else if (colnames[i].equals("pval1")) {
                pvalcol = i;
            } else if (colnames[i].equals("pval3")) {
                pval3col = i;
            } else if (colnames[i].equals("red")) {
                redcol = i;
            } else if (colnames[i].equals("green")) {
                greencol = i;
            } else if (colnames[i].equals("medianofratios")) {
                mediancol = i;
            }
        }       
    }

    public void readFile(String fname) throws IOException, SQLException {
        chipchip.setAutoCommit(false);
        BufferedReader reader = new BufferedReader(new FileReader(fname));
        PreparedStatement insert = chipchip.prepareStatement("insert into rosettaresults(analysis, chromosome, position, ratio, " +
                                                             "x, pval, pval3, red, green, medianofratios) values (" +
                                                             "?,?,?,?,?,?,?,?,?,?)");
        insert.setInt(1,analysisid);
        String headerline = reader.readLine();
        getColumnHeaders(headerline);
        String line;     
        int added = 0;
        int duplicate = 0;
        while ((line = reader.readLine()) != null) {
            String pieces[] = line.split("\\t");
            if (pieces[mediancol].equals("FLAG") ||
                pieces[ratiocol].equals("FLAG") ||
                pieces[xcol].equals("FLAG")) {
                continue;
            }

            insert.setInt(2, chrommap.get(pieces[chromcol]));
            insert.setInt(3, Integer.parseInt(pieces[poscol]));
            insert.setDouble(4, Double.parseDouble(pieces[ratiocol]));
            insert.setDouble(5, Double.parseDouble(pieces[xcol]));
            insert.setDouble(6, Double.parseDouble(pieces[pvalcol]));
            insert.setDouble(7, Double.parseDouble(pieces[pval3col]));
            insert.setDouble(8, Double.parseDouble(pieces[redcol]));
            insert.setDouble(9, Double.parseDouble(pieces[greencol]));
            insert.setDouble(10, Double.parseDouble(pieces[mediancol]));
            try {
                insert.execute();
            } catch (SQLException e) {
                if (e.toString().matches(".*unique.*")) {
                    duplicate++;
                    continue;
                } else {
                    throw e;
                }
            }
            added++;
        }
        chipchip.commit();        
        System.err.println("Added " + added + " values from " + fname);
        System.err.println("Dropped " + duplicate + " apparently duplicate probe positions");
    }

    public static void main(String args[]) throws Exception {
        AddMSP msp = new AddMSP();
        msp.parseArgs(args);
        msp.createAnalysis();
        msp.createChromMap();
        for (String fname : msp.fnames) {
            msp.readFile(fname);
        }
    }
}

File: db-scripts/add_msp.pl
#!/usr/bin/perl

use strict;
use warnings;
use DBI;
use Getopt::Long;
use PSRG::Database;
my $dbhcore = PSRG::Database::handleForRole('core');
my $dbhchip = PSRG::Database::handleForRole('chipchip');


my ($species, $genomeversion, $analysis);

GetOptions("analysis=s"=>\$analysis,
	   "species=s"=>\$species);

($species,$genomeversion) = split(';',$species);
my @results = $dbhcore->selectrow_array("select id from species where name = '$species'");
if (@results == 0) {
  die "Can't find species $species";
}
my $speciesid = $results[0];
@results = $dbhcore->selectrow_array("select id from genome where species = $speciesid and version = '$genomeversion'");
if (@results == 0) {
  die "Can't find genome for $speciesid, $genomeversion";
}
my $genomeid = $results[0];

my ($analysisname, $analysisversion) = split(';',$analysis);
@results = $dbhchip->selectrow_array("select id from rosettaanalysis where name = '$analysisname' and version = '$analysisversion' and species = $speciesid");
if (@results == 0) {
  $dbhchip->do("insert into rosettaanalysis (id, species, name, version, active) values (".PSRG::Database::AIinsertText($dbhchip,'rosettaanalysis_id').", $speciesid, '$analysisname', '$analysisversion',1)");
  @results = $dbhchip->selectrow_array("select ".PSRG::Database::AIfetchValue($dbhchip,'rosettaanalysis_id'));
} else {
  warn "Using existing analysis record";
}
my $analysisid = $results[0];

@results = $dbhchip->selectrow_array("select count(*) from rosettaToGenome where analysis = $analysisid and genome = $genomeid");
if ($results[0] == 0) {
  $dbhchip->do("insert into rosettaToGenome(analysis,genome) values($analysisid,$genomeid)");
}

my %chrmap;
my $sth = $dbhcore->prepare("select name, id from chromosome where genome = $genomeid");
$sth->execute();
while (my @row = $sth->fetchrow_array) {
  $chrmap{$row[0]} = $row[1];
}

my %mspchr;
if ($species eq 'Homo sapiens') {
  %mspchr = (23=>'X',
	     24=>'Y',
	     25=>'mt');
} elsif ($species eq 'Mus musculus') {
  %mspchr = (20=>'X',
	     21=>'Y',
	     22=>'mt');
} elsif ($species eq 'Danio rerio') {
  %mspchr = (26=>'Un',
	     27=>'NA');
} elsif ($species eq 'Drosophila melanogaster') {
  %mspchr = (1=>'2L',
	      2=>'2R',
	      3=>'3L',
	      4=>4,
	      5=>'3R',
	      6=>'X',
	      7=>'Yh');
  foreach (keys %chrmap) {
    next if (/^chr/i);
    $mspchr{"chr${_}"} = $_;
  }
}


$dbhchip->{AutoCommit} = 0;
$dbhchip->{RaiseError} = 1;
$sth = $dbhchip->prepare("insert into rosettaresults(analysis,chromosome,position,ratio,x,pval,pval3,red,green,medianofratios) " .
		     "values ($analysisid,?,?,?,?,?,?,?,?,?)");
my @wantcols = qw(chr pos ratio x pval1 pval3 red green medianofratios);
$wantcols[3] = "x'";

foreach my $filename (@ARGV) {
  print STDERR "Reading $filename\n";
  open(FILE,$filename) or die "Can't open $filename : $!";
  my $header = <FILE>;
  chomp($header);
  my @cols = split(/\t/,$header);
  my %cols = ();
  for (my $i = 0; $i <= $#cols; $i++) {
    $cols{lc($cols[$i])} = $i;
  }
  foreach (@wantcols) {
    die "Can't find column $_ in @cols" unless (defined $cols{$_});
  }
  while (my $line = <FILE>) {
    chomp $line;
    next if ($line =~ /FLAG/);
    my @line = split(/\t/,$line);
    if (exists $mspchr{$line[$cols{chr}]}) {
      $line[$cols{chr}] = $mspchr{$line[$cols{chr}]};
    }
    $line[$cols{chr}] = $chrmap{$line[$cols{chr}]} || (warn "No chrom for $line[$cols{chr}]", next);
#    for the H3 depletion file that stuart had
#    $line[$cols{ratio}] = 1/$line[$cols{ratio}];
    eval {
      $sth->execute(map {$line[$cols{$_}]} @wantcols);
    };
    if ($@) {
      if ($@ !~ /ORA-00001/) {
	die $@;
      }
    }
  }

  $dbhchip->commit();
}
$dbhchip->disconnect();
